uat/custom-components/client-python-paho/src/grpc_client_server/grpc_control_server.py

```diff
@@ -0,0 +1,186 @@
+from grpc_client_server.grpc_generated import mqtt_client_control_pb2_grpc
+from grpc_client_server.grpc_generated.mqtt_client_control_pb2 import (
+    Empty,
+    ShutdownRequest,
+    MqttConnectRequest,
+    MqttConnectReply,
+    MqttCloseRequest,
+    MqttSubscribeRequest,
+    MqttSubscribeReply,
+    MqttUnsubscribeRequest,
+    MqttPublishRequest,
+    MqttPublishReply,
+)
+
+
+class GRPCControlServer(
+    mqtt_client_control_pb2_grpc.MqttClientControlServicer
+):
+    """Implementation of gRPC server handles requests of OTF."""
+
+    logger = logging.getLogger("GRPCControlServer")
+
+    def __init__(self, client: GRPCDiscoveryClient, address: str):
+        """
+        Construct GRPCControlServer
+        Parameters
+        ----------
+        client - GRPCDiscoveryClient object
+        address - local gRPC service address
+        """
+        self.__logger = GRPCControlServer.logger
+        self.__shutdown_reason = ""
+        self.__stop_server_future = asyncio.Future()
+
+        # TODO delete suppress warning
+        self.__client = client  # pylint: disable=unused-private-member
+
+        super().__init__()
+
+        self.__server = grpc.aio.server()
+        mqtt_client_control_pb2_grpc.add_MqttClientControlServicer_to_server(
+            self, self.__server
+        )
+        self.__chosen_port = self.__server.add_insecure_port(address)
+
+    def get_port(self) -> int:
+        """Returns gRPC server port"""
+        return self.__chosen_port
+
+    def get_shutdown_reason(self) -> str:
+        """Returns shutdown reason"""
+        return self.__shutdown_reason
+
+    async def start_grpc_server(self):
+        """Start grpc server"""
+        await self.__server.start()
+
+    async def wait(self):
+        """Wait until incoming shutdown request"""
+        self.__logger.info("Server awaiting termination")
+        await self.__stop_server_future
+        await self.__server.stop(None)
+        await self.__server.wait_for_termination()
+        self.__logger.info("Server termination done")
+
+    def unblock_wait(self):
+        """Unblock wait method"""
+        try:
+            self.__stop_server_future.set_result(True)
+        except asyncio.base_futures.InvalidStateError:
+            pass
+
+    def close(self):
+        """Closes the gRPC server."""
+        self.unblock_wait()
+
+    async def ShutdownAgent(
+        self, request: ShutdownRequest, context: grpc.aio.ServicerContext
+    ) -> Empty:
+        """
+        override
+        Handler of ShutdownAgent gRPC call.
+        Parameters
+        ----------
+        request - incoming request
+        context - request context
+        Returns Empty object
+        """
+        self.__shutdown_reason = request.reason
+        self.__logger.info("shutdownAgent: reason %s", self.__shutdown_reason)
+        self.unblock_wait()
+        return Empty()
+
+    async def CreateMqttConnection(

```

Comment:
why is this capitalized? Python functions should not be capitalized.

Reply:
These capitalized functions are inherited from the GRPC-generated code (`from grpc_client_server.grpc_generated import mqtt_client_control_pb2_grpc`).

Generated code takes methods names from ".proto" file as is.
For example, in ".proto" we have 
`rpc CreateMqttConnection(MqttConnectRequest) returns (MqttConnectReply) {}`

The generated "mqtt_client_control_pb2_grpc.py" file code
`def CreateMqttConnection(self, request, context):`

We cannot change these method names to be aligned with the python code style.
In "grpc_control_server.py" I marked inherited overridden methods from generated code with "override" keyword in the methods headers.